proposals/0000-reconsider-typealias-in-protocol.md

```diff
@@ -0,0 +1,138 @@
+protocol P1: P {associatedtype A}
+// warning: associated type 'R' is redundant with type 'R' declared in inherited protocol 'P'
+```
+#### `#3`
+```swift
+protocol P {
+  associatedtype A
+  func foo() -> A // error: 'A' is ambiguous for type lookup in this context
+}
+extension P {typealias A = Int}
+```
+### Solution
+Now that we have powerful and expressive where clauses for protocol declarations,
+I propose to bring consistency to the usage of type aliases within protocols and protocol extensions.
+
+* [Type aliases in protocol extensions](#type-aliases-in-protocol-extensions). There are two options:
+    * Permit type aliases to act as default values for associated types in protocol extensions. This is the preferred           solution. 
+    * Change the error message to propose the current syntax instead of simply producing **#3**,
+    which is actually a generic error.
+* Type aliases in protocols should be always discouraged in **Swift 4.2** and always disallowed in **Swift 5**, the same way we disallow any implementation details.
+* Warning **#1** should be amended to a generic `typealias should/must not be declared within a protocol` warning/error for **Swift 4.2** and **Swift 5** respectively with fix-it options to either switch to an `associatedtype` or impose a same-type constraint on the protocol.
+* Warning **#2** should be superseded by the `typealias should/must not be declared within a protocol` warning/error in **Swift 4.2** and **Swift 5** respectively.
+
+## Detailed design
+
+#### Non-conflicting type alias in protocol
+```swift
+protocol P {
+ typealias A = Int // fixit: typealias -> associatedtype
+ /// Swift 4.2
+ /// warning: type alias should not be declared within a protocol;
+ /// move to an extension or use 'associatedtype' to define an associated type requirement 
+ /// Swift 5
+ /// error: type alias must not be declared within a protocol;
+ /// move to an extension or use 'associatedtype' to define an associated type requirement
+}
+```
+#### Type alias conflicting with inherited associated type
+
+```swift
+protocol P {associatedtype A}
+protocol P1: P {    // fix-it: where A == Int
+  typealias A = Int // fix-it: typealias -> associatedtype
+  /// Swift 5
+  /// error: typealias must not be declared within a protocol
+  /// note: use 'associatedtype' to specify a default value for 'A' from protocol 'P'
+  /// note: use a same-type constraint on the protocol
+}
+```
+#### Type aliases in protocol extensions
+The benign second options stated under [Solution](#solution) will not break source, but will remain a contradictory case on the background of the proposed changes and in constrast to established language rules.
+
+The first case, however, is just as harmless: switching from same-type constraints to default values is source breaking **neither** in the case when there are no appliances of the `associatedtype` among the requirements, nor when there are: it relaxes an existing error.
```

Comment:
Hmm, this isn't true, is it? The following example currently (in Swift 4.2) warns about the type alias in the protocol `Q`, then treats it as a same-type constraint. Generic functions can rely on that same-type constraint:

```swift
protocol P {
  associatedtype E
  func getE() -> Self.E
 }

protocol Q: P {
  typealias E = Bool
}

func foo<T: Q>(_ t: T) -> Bool {
  return t.getE() // okay: we know that E == Bool
}
```

We can still stage in the proposed changes in Swift 5, and maintain this behavior for Swift < 5, but the proposal should make it clear that this is a source-breaking change (albeit a migratable one, with some effort).

Reply:
Most of the text above belongs in the "Source Compatibility" section?